compiler/shared/main/scala/mlscript/compiler/ir/Builder.scala

```diff
@@ -60,64 +97,212 @@ final class Builder(fresh: Fresh, fnUid: FreshInt, classUid: FreshInt, tag: Fres
     }
     tm
 
-  private def buildResultFromTerm(using ctx: Ctx)(tm: Term)(k: Node => Node): Node =
-    def buildLetCall(f: Term, xs: Tup, ann: Option[Term]) =
-      buildResultFromTerm(f) { node => node match
-        case Result(Ref(g) :: Nil) if ctx.fnCtx.contains(g.str) => buildResultFromTerm(xs) {
-          case Result(args) =>
-            val v = fresh.make
-            
-            ann match
-              case Some(ann @ Var(nme)) =>
-                if nme === "tailcall" then 
-                  LetCall(List(v), DefnRef(Right(g.str)), args, true, v |> ref |> sresult |> k)(f.toLoc).attachTag(tag)
-                else
-                  if nme === "tailrec" then
-                    raise(ErrorReport(List(msg"@tailrec is for annotating functions; try @tailcall instead" -> ann.toLoc), true, Diagnostic.Compilation))
-                  LetCall(List(v), DefnRef(Right(g.str)), args, false, v |> ref |> sresult |> k)(f.toLoc).attachTag(tag) 
-              case Some(_) => node |> unexpectedNode
-              case None => LetCall(List(v), DefnRef(Right(g.str)), args, false, v |> ref |> sresult |> k)(f.toLoc).attachTag(tag)   
-            
-          case node @ _ => node |> unexpectedNode
-        }
-        case Result(Ref(f) :: Nil) => buildResultFromTerm(xs) {
-          case Result(args) =>
-            throw IRError(s"not supported: apply")
-          case node @ _ => node |> unexpectedNode
+  private def freeVariablesIf(ucs: If): Set[Str] =
+    val ifbody: IfBody = ucs.body
+    val els: Opt[Term] = ucs.els
+
+    def f(ifbody: IfBody): Set[Str] = ifbody match
+      case IfBlock(lines) => lines.foldLeft(Set.empty[Str]) {
+        case (acc, Left(ifbody2)) => acc ++ f(ifbody2)
+        case (acc, Right(rhs)) => acc ++ freeVariables(rhs)._2
+      }
+      case IfElse(expr) => freeVariables(expr)
+      case IfLet(isRec, name, rhs, body) =>
+        if isRec then
+          (freeVariables(rhs) -- freeVariables(name)) ++ (f(body) -- freeVariables(name))
+        else
+          freeVariables(rhs) ++ (f(body) -- freeVariables(name))
+      case IfOpApp(lhs, op, rhs) => freeVariables(lhs) ++ f(rhs)
+      case IfOpsApp(lhs, opsRhss) => freeVariables(lhs) ++ opsRhss.foldLeft(Set.empty[Str]) {
+        case (acc, (op, rhs)) => acc ++ f(rhs)
+      }
+      case IfThen(expr, rhs) => freeVariables(rhs) -- freeVariables(expr)
+
+    val fvs1 = f(ifbody)
+    val fvs2 = els.fold(Set.empty[Str]) { freeVariables }
+
+    fvs1 ++ fvs2
+
+  private def freeVariables(tu: TypingUnit): Set[Str] =
+    var all_defined = tu.rawEntities.foldLeft(Set.empty[Str]) {
+      case (defined, stmt) => defined ++ freeVariables(stmt)._1
+    }
+    tu.rawEntities.foldLeft(Set.empty[Str]) {
+      case (acc, stmt) => acc ++ (freeVariables(stmt)._2 -- all_defined)
+    }
+
+    // for this fv analysis, we also need to return the variables that are defined in this statement
+  private def freeVariables(stmt: Statement): (Set[Str], Set[Str]) = stmt match
```

Comment:
Any reason you're not reusing or at least adapting/fixing this one?

https://github.com/hkust-taco/mlscript/blob/4a5a0387a6bd12c93de153a6c510f868f7342aa2/shared/src/main/scala/mlscript/helpers.scala#L895

Reply:
Ok I see your version returns two sets. Is that really necessary?

Reply:
Oops, never noticed it. I will try to see if it could replace the implementation in the IR builder.

Reply:
I tried this implementation, first, it will recognize `bar` as one of the free variables of `foo.bar`. It will also think the operator `is` of ucs is a free variable.